DND Override functionality for navigator

PhilMFischer opened the pull request:
- [x] Added extension point to navigator to provide concept based overrides for DragAndDrop
- [x] Refactored navigator to deal with provided drop adapters
- [x] Still need to write unit tests
- [x] May provide swtbot tests for drag and drop

Refactored AProjectTestCase to have some more simple API to setup test cases. Adjusted all test cases in project to comply with these changes now.

SWTBOT tests only provided for Drag And Drop Move operations. Copy and Link not yet supported by SWTBOT API.

Commit:
Added API for delegated drop adapters - (Task #74)

- Added drop adapter to navigator which is capable of redirecting drag
and drop operations to different drop handlers.
- Added extension point to register new drop adapters for individual
concepts.

---
Task #74: Override DND functionality in Navigator

Commit:
Added Empty Implementation for PS Drop Adapter - (Task #74)

- Added concept specific dropa dapter to Product structure model
extension

---
Task #74: Override DND functionality in Navigator

Commit:
Merge remote-tracking branch 'origin/development' into feature/task_74-Override_DND_functiona

Commit:
Implemented first Product Structure related dropAdapter - (Task #74)

- Renamed extension point schema.
- Fixed code to load delegation drop adapter from registration
- Implemented first drop adapter to replace typing of a given
structuralelementinstance.

---
Task #74: Override DND functionality in Navigator

Commit:
Refactored Command Checks - (Task #74)

- Now checking if object can be inherited from, otherwise a command will
be set to be unexecutable.

---
Task #74: Override DND functionality in Navigator

SaMuellerDLR committed:
Added test case for inheritance case - (Task #74)

- Added test case to test that the applicability check is correct when
using the inheritance feature
- Renamed test cases to distinguish between the cases: adding child seis
or adding super seis.

---
Task #74: Override DND functionality in Navigator

Commit:
moved implementation for typing to project - (Task #74)

- Inheritance is in general an extension independant functionality,
therefore it has been moved to project.
- will implement PS specific inheritance drop to also create new
elements.

---
Task #74: Override DND functionality in Navigator

Commit:
Merge branch 'feature/task_74-Override_DND_functiona' of https://github.com/virtualsatellite/VirtualSatellite4-Core.git into feature/task_74-Override_DND_functiona

Commit:
Implemented all functionality for PS drag and drop - (Task #74)

- Added functionality to create new PS element when pressing ALT with
DND operation.
- Also adjusting discipline and names when creating drop.

---
Task #74: Override DND functionality in Navigator

Commit:
Cleaned up code - (Task #74)

- Removed all warnings
- Prepared test cases

---
Task #74: Override DND functionality in Navigator

SaMuellerDLR commented on de.dlr.sc.virsat.model.extension.ps/src/de/dlr/sc/virsat/model/extension/ps/dnd/ProductStructureDragAndDropInheritanceCommandHelper.java:
```diff
@@ -0,0 +1,165 @@
+			}
+		} catch (CoreException | NoSuchMethodException | SecurityException | InstantiationException | IllegalAccessException | IllegalArgumentException | InvocationTargetException e) {
+			Activator.getDefault().getLog().log(new Status(
+					Status.INFO,
```
Is this really just an info and not an error?

PhilMFischer replied:
Well a WARNING could be ok, ERROR I think should indicate a problem that may create havoc when further executing the application. WARNING is more like something went unexpectedly wrong but is regarded as no critical problem.

SaMuellerDLR commented on de.dlr.sc.virsat.model.extension.ps/src/de/dlr/sc/virsat/model/extension/ps/dnd/ProductStructureDragAndDropInheritanceCommandHelper.java:
```diff
@@ -0,0 +1,165 @@
+	 * @param ed The editing domain in which the object will be dropped
+	 * @param dragObjects The objects preferably SEIs which are dragged
+	 * @param dropSei the SEI to which the objects shall get dropped
+	 * @return A BeanSEI with corresponding to the drop target with correct name and discipline
```
The first with can be removed

Commit:
Write a Heading here - (Task #74)

- Added further hamcrest dependency to test plugin so that the
hamcrest.collections matchers are available.
- Implemented test cases for VirSatInheritance Drop Adapter

---
Task #74: Override DND functionality in Navigator

SaMuellerDLR commented on de.dlr.sc.virsat.project.ui/src/de/dlr/sc/virsat/project/ui/navigator/dropAssist/DVLMObjectDropAdapterAssistant.java:
```diff
@@ -9,98 +9,176 @@
+						Activator.getDefault().getLog().log(new Status(Status.ERROR, Activator.getPluginId(), "Cannot register two drop adapter for the same concept: " + conceptId));
+					}
+				} catch (CoreException e) {
+					Activator.getDefault().getLog().log(new Status(Status.ERROR, Activator.getPluginId(), "Failed to create executable extension for drop adapter in concept: " + navigatorExtension.getContributor() + "->" + conceptId + " because: " + e.getMessage()));
```
I think we should also pass the stack trace to the logger.

PhilMFischer replied:
Yes added it to the places where INFOs are now logged as WARNINGS

Commit:
Simplified concept test case setup - (Task #74)

- Added method to easily execute tests as a command in the editing
domain.

- Removed some further warnings.

---
Task #74: Override DND functionality in Navigator

Commit:
Refactored execute function in project test case  - (Task #74)

- Added a second method to execute code as recording command in the ED.
But this time the method does not hand back results.

---
Task #74: Override DND functionality in Navigator

Commit:
Implemented Test cases for PS inheritance commands - (Task #74)

- Refatcored command helper to mainly use PS bean elements where
possible
- Added test cases

---
Task #74: Override DND functionality in Navigator

Commit:
Refactored all project tests - (Task #74)

- Refactored due to changes in AProjectTest case
- Unified setup and tear down of all project unit tests.

---
Task #74: Override DND functionality in Navigator

Commit:
Merge remote-tracking branch 'origin/feature/task_430-Use_https_based_addre' into feature/task_74-Override_DND_functiona

Commit:
Improved SWT Bot tests with drag and drop - (Task #74)

- Added DND testcase for moving a SEI
- Added DND testcase for moving a CA
- Added DND testcase for setting inheritance
- Added hamcrest library to dependencies for better hamcrest assert
statements on collections etc.

---
Task #74: Override DND functionality in Navigator

Commit:
Minor fixes due to review - (Task #74)

- changed some log statements from info to warning level.

---
Task #74: Override DND functionality in Navigator

Commit:
Fixing test case on jenkins - (Task #74)

---
Task #74: Override DND functionality in Navigator

Commit:
Integrated old test case - (Task #74)

- Old test case for project builder was not added to all tests.

---
Task #74: Override DND functionality in Navigator

Commit:
Fixing SWT Bot dependencies - (Task #74)

- Added all concept plugins.

---
Task #74: Override DND functionality in Navigator

Commit:
Merge remote-tracking branch 'origin/development' into feature/task_74-Override_DND_functiona

SaMuellerDLR commented on de.dlr.sc.virsat.model.extension.ps.test/src/de/dlr/sc/virsat/model/extension/ps/util/test/ProductStructureHelperTest.java:
```diff
@@ -37,14 +33,12 @@
+	
+	private static final int AMOUNT_CT_SUBSYTEMS = 3;
 
 	ElementDefinition ed;
```
I think we can make these private.

PhilMFischer replied:
arent they automatically private if not explicitly stated?

SaMuellerDLR replied:
The visibility level if not stated is considered "package protected". E.g. classes in the same package can access the variables. https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html 

This has the minor consequence that eclipse doesnt create warnings for unused variables. So if, for example due to a change in the test case, the variable ed is no longer used, then eclipse would not create a warning. And the dead variable would just lie around.

PhilMFischer replied:
ok, than i add it to the change.

SaMuellerDLR commented on de.dlr.sc.virsat.project.test/src/de/dlr/sc/virsat/project/structure/VirSatProjectCommonsTest.java:
```diff
@@ -65,6 +65,10 @@
 	
+	@Override
+	protected void addProjectFileStructure() {
+	}
```
Maybe a short comment why its overwritten.

PhilMFischer replied:
Done

SaMuellerDLR commented on de.dlr.sc.virsat.project.ui/src/de/dlr/sc/virsat/project/ui/navigator/dropAssist/ADVLMDropAdapaterAssistant.java:
```diff
@@ -0,0 +1,86 @@
+public abstract class ADVLMDropAdapaterAssistant extends CommonDropAdapterAssistant {
+
+	public ADVLMDropAdapaterAssistant() {
+		super();
```
I dont think we need an empty constructor.

PhilMFischer replied:
Done

SaMuellerDLR commented on de.dlr.sc.virsat.swtbot.test/src/de/dlr/sc/virsat/swtbot/test/InheritanceTest.java:
```diff
@@ -139,6 +140,40 @@ public void singleLevelInheritancePropagationUITest() {
+		waitForAllBuildersAndUiThread();
 		
+		openEditor(elementOccurence);
+		//waitForAllBuildersAndUiThread();
```
Do we need this commented out code?

PhilMFischer replied:
Removed

SaMuellerDLR commented on de.dlr.sc.virsat.project.test/src/de/dlr/sc/virsat/project/test/AProjectTestCase.java:
```diff
@@ -160,4 +164,47 @@ protected void addEditingDomainAndRepository() {
+	protected <T> T executeAsCommand(Supplier<T> supplier) {
+		
+		// Simple class to store object of the given generic type T
+		class Store {
```
I guess this Store class is introduced to avoid the effectivly final requirement since you want to save the result from the doExecute call?
You may disagree, but I would prefer using `final List<T> objects = new ArrayList<>();` and then `objects.add(supplier.get())` simple since it avoids introducing some new magic class.

PhilMFischer replied:
Decided to go with a vector

Commit:
Fixes due to review comments - (Task #74)

- Major change, using vector to store object from the execute method
lambda expression.

---
Task #74: Override DND functionality in Navigator

Commit:
Added some comments as requested by Review - (Task #74)

---
Task #74: Override DND functionality in Navigator

Commit:
Fixes due to review - (Task #74)

- made variables private and removed them due to "unused"

---
Task #74: Override DND functionality in Navigator

SaMuellerDLR reviewed (approved):
Approved! Great work.

andriikovalov-dlr commented on de.dlr.sc.virsat.swtbot.test/META-INF/MANIFEST.MF:
```diff
@@ -43,7 +43,20 @@ Require-Bundle: org.eclipse.swt,
- de.dlr.sc.virsat.model.extension.budget.mass.ui
+ de.dlr.sc.virsat.model.extension.budget.mass.ui,
+ org.hamcrest.library,
+ de.dlr.sc.virsat.model.extension.budget.power,
```
Why do we need to add all these dependencies? I don't see them used in the tests

PhilMFischer replied:
Good question. SWTBot seems to start up the VirSat/Eclipse application differently then we do it. Our application is usually based on features. For SWTBot even specifiying the launch configuration the same way, seems to do things differently. If these dependencies are not explicitly added, Virtual Satellite is not executed correctly. 

For example, all concepts will be discovered and displayed in the concept palette. The according images will not be shown because the plugins containing the icons cannot be found/loaded.

Adding the dependencies all images and icons are found as exepcted.

andriikovalov-dlr commented on de.dlr.sc.virsat.swtbot.test/src/de/dlr/sc/virsat/swtbot/test/InheritanceTest.java:
```diff
@@ -139,6 +140,38 @@ public void singleLevelInheritancePropagationUITest() {
+		waitForAllBuildersAndUiThread();
 		
+		openEditor(elementOccurence);
+		assertEquals("Received override from EC2", "docFromEc2", bot.tableWithId("tableDocument").cell(0, Document.PROPERTY_DOCUMENTNAME));
```
Let's maybe also check the number of child items after DND?
Also can we emulate add and override actions?

PhilMFischer replied:
1. Well, we could check for the children, but i think that is covered by the unit tests we added. This test i just implemented to have a rough overview that the registration on the extension point and execution of the dropAdapter works as expected.
2. Unfortunately SWTBot only supports The DragAndDrop Move at the moment. Changing it looks to me like a modification in SWTBot code.

PhilMFischer commented on de.dlr.sc.virsat.project.test/src/de/dlr/sc/virsat/project/editingDomain/commands/VirSatClipboardCommandHelperTest.java:
```diff
@@ -31,19 +29,9 @@
- * @author fisc_ph
- *
- */
+W */
```
@andriikovalov-dlr Cool thing, how does this work with the suggestion 🤓

Commit:
Update de.dlr.sc.virsat.project.test/src/de/dlr/sc/virsat/project/editingDomain/commands/VirSatClipboardCommandHelperTest.java

Co-Authored-By: andriikovalov-dlr <48556822+andriikovalov-dlr@users.noreply.github.com>

Commit:
Update de.dlr.sc.virsat.project.ui/src/de/dlr/sc/virsat/project/ui/navigator/dropAssist/ADVLMDropAdapaterAssistant.java

Co-Authored-By: andriikovalov-dlr <48556822+andriikovalov-dlr@users.noreply.github.com>

PhilMFischer commented on de.dlr.sc.virsat.project.ui/src/de/dlr/sc/virsat/project/ui/navigator/dropAssist/ADVLMDropAdapaterAssistant.java:
```diff
@@ -0,0 +1,82 @@
+			validatedDragObjects = getSelectedDragObjects();
+			validatedDropObject = (EObject) dropObject;
+			ed = VirSatEditingDomainRegistry.INSTANCE.getEd((EObject) dropObject);
+			validatedDropCommand = createDropCommand(
```
Yes, i am not happy with it either. The way it is written it is readable, autoformatting will fail big time. Is there a good way we could adjust our formatting?

PhilMFischer commented on de.dlr.sc.virsat.model.edit/src/de/dlr/sc/virsat/model/dvlm/util/DVLMCommandParameterApplicableForCheck.java:
```diff
@@ -9,22 +9,34 @@
+
 
 /**
  * This class extends the ApplicableFor checks to be used with EMF CommandParameters
```
Got adjusted.

PhilMFischer commented on de.dlr.sc.virsat.project.ui/src/de/dlr/sc/virsat/project/ui/navigator/dropAssist/ADVLMDropAdapaterAssistant.java:
```diff
@@ -0,0 +1,82 @@
+	@Override
+	public IStatus validateDrop(Object dropObject, int operation, TransferData transferType) {
+		if (dropObject instanceof EObject) {
+			validatedDragObjects = getSelectedDragObjects();
```
Not a hundred percent sure if i ilke to name it differently, because for me it were the objects which were ok when the validateDrop method executed. Nevertheless, they are not necessarily valid. I changed them now to dropEObject and dragObjects

PhilMFischer commented on de.dlr.sc.virsat.project.ui/src/de/dlr/sc/virsat/project/ui/navigator/dropAssist/DVLMObjectDropAdapterAssistant.java:
```diff
@@ -9,98 +9,176 @@
- *
+ * Drop Assistant for VirSat Navigator that delegates calls to concept specific drop assistants
  */
 public class DVLMObjectDropAdapterAssistant extends CommonDropAdapterAssistant {
```
> This shoud be named somehow like DelegatedDropAdapterAssistant

Done.

PhilMFischer replied:
> To me it looks like this class mostly repeats platform code. Can't we just use the normal navigator dropassistant extension point from concepts? We can make them higher priority to be sure they act before the default ones

Unfortunately no. From what i saw, we can only add dropAdapter to navigator content contributions. Therefore every concept may have to provide a complete navigator content. Not a 100% sure, but i also think that it is only possible to add one dropAdapter per navigatorContent.

The way it is implemented now, it is possible to define a dropAdapter which can be used in any navigator but it can also be reused in the delegateDropAdapter.

I think this increases functionality and flexibility on what developers want to do on concept side.

PhilMFischer commented on de.dlr.sc.virsat.project.ui/src/de/dlr/sc/virsat/project/ui/navigator/dropAssist/DVLMObjectDropAdapterAssistant.java:
```diff
@@ -9,98 +9,176 @@
- *
+ * Drop Assistant for VirSat Navigator that delegates calls to concept specific drop assistants
  */
 public class DVLMObjectDropAdapterAssistant extends CommonDropAdapterAssistant {
```
Good question! @SaMuellerDLR  Do we need the drop adapter there? I think it is heritage and we should leave it there for the moment. But i guess we don't need it , since we do not have any code dealing with Copy&Paste and DnD of Equations. Do we?

SaMuellerDLR replied:
I checked it yesterday and we dont have any drag and drop functionality for equations.
I found the commit in our old repository, and there it looks like the entry only exists due to creating the setup with copy & paste. I think it can be safely removed.

andriikovalov-dlr replied:
I removed it from calculations

PhilMFischer commented:
@andriikovalov-dlr and @SaMuellerDLR Please review the comments and close them if you agree.

Commit:
Changes due to review of andrii - (Task #74)

- fixed some comments
- renamed class
- renamed some member variables

---
Task #74: Override DND functionality in Navigator

Commit:
Merge with changes from GitHub Review - (Task #74)


---
Task #74: Override DND functionality in Navigator

andriikovalov-dlr committed:
Removed unused dropAssistant from calculations - (Task #74)

---
Task #74: Override DND functionality in Navigator
